### PR TITLE
refactor(types): adjust Subscription and Observable types to be easier to consume externally

### PR DESCRIPTION
**Pull request by jasonaden**
This PR removes the internal pieces of `Subscription` and `Observable` as well as consolidates interfaces and types to `types.ts`. Some of the interface names were adjusted as well to be more usable externally from Rx.

### CODE REVIEW

File: src/internal/operators/onErrorResumeNext.ts [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+import { Observable } from '../Observable';
 import { from } from '../observable/from';
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { isArray } from '..//util/isArray';
 import { OuterSubscriber } from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import { subscribeToResult } from '..//util/subscribeToResult';
-import { OperatorFunction } from '../../internal/types';
+import { ObservableInput, OperatorFunction } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/onErrorResumeNext.ts]: rename import to ../types to be consistent

File: src/internal/operators/refCount.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/refCount.ts]: merge types imports together.

File: src/internal/operators/repeat.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
 import { empty } from '../observable/empty';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/repeat.ts]: merge types imports together.

File: src/internal/operators/repeatWhen.ts [hunk 4]
```diff
@@ -2,7 +2,8 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
 import { Subject } from '../Subject';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/repeatWhen.ts]: merge types imports together.

File: src/internal/operators/retry.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/retry.ts]: merge types imports together.

File: src/internal/operators/retryWhen.ts [hunk 6]
```diff
@@ -2,7 +2,8 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
 import { Subject } from '../Subject';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/retryWhen.ts]: merge types imports together.

File: src/internal/operators/sample.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/sample.ts]: merge types imports together.

File: src/internal/operators/sampleTime.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { Subscriber } from '../Subscriber';
 import { IScheduler } from '../Scheduler';
 import { Action } from '../scheduler/Action';
 import { async } from '../scheduler/async';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/sampleTime.ts]: merge types imports together.

File: src/internal/operators/sequenceEqual.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
 import { Operator } from '../Operator';
-import { Observer } from '../Observer';
+import { Observer } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/sequenceEqual.ts]: merge types imports together.

File: src/internal/operators/single.ts [hunk 10]
```diff
@@ -1,9 +1,9 @@
 import { Observable } from '../Observable';
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
-import { Observer } from '../Observer';
+import { Observer } from '../types';
 import { EmptyError } from '..//util/EmptyError';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/single.ts]: merge types imports together.

File: src/internal/operators/skip.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/skip.ts]: merge types imports together.

File: src/internal/operators/skipLast.ts [hunk 12]
```diff
@@ -2,7 +2,7 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { ArgumentOutOfRangeError } from '..//util/ArgumentOutOfRangeError';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/skipLast.ts]: merge types imports together.

File: src/internal/operators/skipUntil.ts [hunk 13]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/skipUntil.ts]: merge types imports together.

File: src/internal/operators/skipWhile.ts [hunk 14]
```diff
@@ -1,7 +1,7 @@
 import { Observable } from '../Observable';
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/skipWhile.ts]: merge types imports together.

File: src/internal/operators/subscribeOn.ts [hunk 15]
```diff
@@ -2,7 +2,7 @@ import { Operator } from '../Operator';
 import { IScheduler } from '../Scheduler';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/subscribeOn.ts]: merge types imports together.

File: src/internal/operators/switchAll.ts [hunk 16]
```diff
@@ -1,5 +1,4 @@
-import { OperatorFunction } from '../../internal/types';
-import { ObservableInput } from '../Observable';
+import { OperatorFunction, ObservableInput } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/switchAll.ts]: rename import to ../types to be consistent

File: src/internal/operators/switchMap.ts [hunk 17]
```diff
@@ -1,11 +1,11 @@
 import { Operator } from '../Operator';
-import { Observable, ObservableInput } from '../Observable';
+import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
 import { Subscription } from '../Subscription';
 import { OuterSubscriber } from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import { subscribeToResult } from '..//util/subscribeToResult';
-import { OperatorFunction } from '../../internal/types';
+import { ObservableInput, OperatorFunction } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/switchMap.ts]: rename import to ../types to be consistent

File: src/internal/operators/switchMapTo.ts [hunk 18]
```diff
@@ -1,11 +1,11 @@
 import { Operator } from '../Operator';
-import { Observable, ObservableInput } from '../Observable';
+import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
 import { Subscription } from '../Subscription';
 import { OuterSubscriber } from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import { subscribeToResult } from '..//util/subscribeToResult';
-import { OperatorFunction } from '../../internal/types';
+import { ObservableInput, OperatorFunction } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/switchMapTo.ts]: rename import to ../types to be consistent

File: src/internal/operators/take.ts [hunk 19]
```diff
@@ -3,7 +3,7 @@ import { Subscriber } from '../Subscriber';
 import { ArgumentOutOfRangeError } from '..//util/ArgumentOutOfRangeError';
 import { empty } from '../observable/empty';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/take.ts]: merge types imports together.

File: src/internal/operators/takeLast.ts [hunk 20]
```diff
@@ -3,7 +3,7 @@ import { Subscriber } from '../Subscriber';
 import { ArgumentOutOfRangeError } from '..//util/ArgumentOutOfRangeError';
 import { empty } from '../observable/empty';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/takeLast.ts]: merge types imports together.

File: src/internal/operators/takeUntil.ts [hunk 21]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/takeUntil.ts]: merge types imports together.

File: src/internal/operators/takeWhile.ts [hunk 22]
```diff
@@ -1,7 +1,7 @@
 import { Operator } from '../Operator';
 import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/takeWhile.ts]: merge types imports together.

File: src/internal/operators/tap.ts [hunk 23]
```diff
@@ -1,8 +1,8 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { Observable } from '../Observable';
-import { PartialObserver } from '../Observer';
-import { TeardownLogic } from '../Subscription';
+import { PartialObserver } from '../types';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/tap.ts]: merge types imports together.

File: src/internal/operators/throttle.ts [hunk 24]
```diff
@@ -1,7 +1,9 @@
 import { Operator } from '../Operator';
-import { Observable, SubscribableOrPromise } from '../Observable';
+import { Observable } from '../Observable';
+import { SubscribableOrPromise } from '../types';
 import { Subscriber } from '../Subscriber';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/throttle.ts]: merge types imports together.

File: src/internal/operators/throttleTime.ts [hunk 25]
```diff
@@ -1,7 +1,8 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { IScheduler } from '../Scheduler';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/throttleTime.ts]: merge types imports together.

File: src/internal/operators/timeout.ts [hunk 26]
```diff
@@ -5,7 +5,7 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { IScheduler } from '../Scheduler';
 import { Observable } from '../Observable';
-import { TeardownLogic } from '../Subscription';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/timeout.ts]: merge types imports together.

File: src/internal/operators/timeoutWith.ts [hunk 27]
```diff
@@ -3,12 +3,12 @@ import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
 import { IScheduler } from '../Scheduler';
 import { async } from '../scheduler/async';
-import { TeardownLogic } from '../Subscription';
-import { Observable, ObservableInput } from '../Observable';
+import { TeardownLogic } from '../types';
```
Comment (david-driscoll) [inline src/internal/operators/timeoutWith.ts]: merge types imports together.

File: src/internal/operators/withLatestFrom.ts [hunk 28]
```diff
@@ -1,10 +1,10 @@
 import { Operator } from '../Operator';
 import { Subscriber } from '../Subscriber';
-import { Observable, ObservableInput } from '../Observable';
+import { Observable } from '../Observable';
 import { OuterSubscriber } from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import { subscribeToResult } from '..//util/subscribeToResult';
-import { OperatorFunction } from '../../internal/types';
+import { ObservableInput, OperatorFunction } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/withLatestFrom.ts]: merge types imports together.

File: src/internal/operators/zipAll.ts [hunk 29]
```diff
@@ -1,6 +1,6 @@
 import { ZipOperator } from '../observable/zip';
-import { Observable, ObservableInput } from '../Observable';
-import { OperatorFunction } from '../../internal/types';
+import { Observable } from '../Observable';
+import { OperatorFunction, ObservableInput } from '../../internal/types';
```
Comment (david-driscoll) [inline src/internal/operators/zipAll.ts]: rename import to ../types to be consistent

File: src/internal/types.ts [hunk 30]
```diff
@@ -10,15 +11,66 @@ export type FactoryOrValue<T> = T | (() => T);
...
+            complete?: () => void): Unsubscribable;
+}
+
 export type ObservableInput<T> = SubscribableOrPromise<T> | ArrayLike<T> | Iterable<T>;
+export type ObservableLike<T> = { [Symbol.observable]: () => Subscribable<T>; };
+
+/** OBSERVER INTERFACES */
+
+export interface NextObserver<T> {
+  closed?: boolean;
```
Comment (david-driscoll) [inline src/internal/types.ts]: @benlesh @kwonoj @mattpodwysocki should `closed` be on the public interface?

File: tsconfig.json [hunk 31]
```diff
@@ -1,28 +1,29 @@
...
+    "removeComments": false,
+    "preserveConstEnums": true,
+    "sourceMap": true,
+    "strictFunctionTypes": true,
+    "noImplicitAny": true,
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "suppressImplicitAnyIndexErrors": true,
+    "moduleResolution": "node",
+    "stripInternal": true,
```
Comment (david-driscoll) [inline tsconfig.json]: For anyone that can't spot the change, `stripInternal` was added here.

File didn't diff correctly, most likely due to the `lf` / `crlf` issue we've seen in a few other files.

File: src/internal/observable/ConnectableObservable.ts [hunk 32]
```diff
@@ -2,7 +2,8 @@ import { Subject, SubjectSubscriber } from '../Subject';
 import { Operator } from '../Operator';
 import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
-import { Subscription, TeardownLogic } from '../Subscription';
+import { Subscription } from '../Subscription';
+import { TeardownLogic } from '../types';
 import { refCount as higherOrderRefCount } from '../../internal/operators/refCount';
 
 /**
```
Comment (david-driscoll) [inline src/internal/observable/ConnectableObservable.ts]: `_isComplete` should be marked `internal` or `protected`.

**Comment by jasonaden** [comment]
@david-driscoll FYI, these were all addressed. I also added a `tslint` rule so duplicate imports will be flagged in the future.

File: src/internal/types.ts (hunk 30 above)
Comment (jasonaden) [reply src/internal/types.ts]: I believe this is closest to the `SubscriptionObserver` interface defined here:

https://github.com/tc39/proposal-observable#subscriptionobserver

Based on how `PartialObserver` is used, I think exposing `closed` here is okay.

File: src/internal/Observable.ts [hunk 33]
```diff
@@ -1,17 +1,14 @@
...
 import { Subscriber } from './Subscriber';
-import { Subscription, AnonymousSubscription, TeardownLogic } from './Subscription';
+import { Subscription } from './Subscription';
+import { TeardownLogic } from './types';
 import { root } from './util/root';
 import { toSubscriber } from './util/toSubscriber';
 import { IfObservable } from './observable/IfObservable';
 import { observable as Symbol_observable } from '../internal/symbol/observable';
-import { OperatorFunction, Subscribable } from '../internal/types';
+import { OperatorFunction, PartialObserver, Subscribable } from '../internal/types';
```
Comment (david-driscoll) [inline src/internal/Observable.ts]: These are actually the same reference, and should be merged.

`../internal/types' ~= `./types' (we're in the internal folder)

**Pull request merged**